### PR TITLE
Run PHP task on all files when theme slug is changed in themeConfig.js.

### PR DESCRIPTION
**Pull request by mor10**
## Description

Addresses issue #30. 

Adds new `watch` task monitoring `themeConfig.js` for changes and runs PHP task any time the file is updated. Compares stored theme slug with theme slug in current `themeConfig.js`, and if no match is found, skips over the `newer` filter forcing all PHP files to be reprocessed so function names are rewritten with the new slug.

## List of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] This pull request relates to a ticket.
- [x] My code is tested.
- [x] I want my code added to WP Rig.

### CODE REVIEW

**Comment by hellofromtonya** [comment]
Using the code as is in this branch, we do not rebuild the PHP files on the first pass when the files already exist and there is one or more changes in the `themeConfig.js` file.  It works in watch mode though.

How to reproduce?

1. Do `npm run build`.
2. Then stop the watcher.
3. Change the theme's slug in `themeConfig.js`.
4. Do `npm run build`.  **Result:  PHP Files are not rebuilt.**

Whoopsie.  That doesn't fit all scenarios.

## Altering the Code

When removing the new code in the `php()` function as well as the original `.pipe(newer(paths.php.dest))`, bam, it works in all scenarios.  WooHoo!

### Test Results

Here are the tests I ran:

1. Changed the theme's slug in the `themeConfig.js`.  Then ran `npm run build`.  **Result: Works.**
2. Changed the theme's slug while in watch. **Result: Works.**
3. Changed the theme's name while in watch mode. **Result: Works.**
4. Stopped the watcher.  Changed both the theme's slug and name. Restarted. **Result: Works.**
5. Repeated 2 and 3. **Result: Works.**
6. Stopped. Changed the name only. Restarted. **Result: Works.**
7. Repeated 2 and 3. **Result: Works.**
8. Repeated the process of changing slug, name, or both.  **Result: Works.**

### New Code

@mor10 Here's the new code for the `php` function.

```
/**
 * Build the PHP files and run PHP Code Sniffer.
 */
export function php() {
	config = requireUncached('./dev/config/themeConfig.js');

	return gulp.src(paths.php.src)
	// Run PHPCS.
	.pipe(phpcs({
		bin: 'vendor/bin/phpcs',
		standard: 'WordPress',
		warningSeverity: 0
	}))
	// Log all problems that was found
	.pipe(phpcs.reporter('log'))

	// Replace the theme's slug and name in the build files.
	.pipe(replace('wprig', config.theme.slug))
	.pipe(replace('WP Rig', config.theme.name))

	// Set the build destinations.
	.pipe(gulp.dest(paths.verbose))
	.pipe(gulp.dest(paths.php.dest));
}
```

**Comment by hellofromtonya** [comment]
@mor10 Here's the revised [`gulpfile.babel.js` in my copy of the branch](https://github.com/hellofromtonya/wprig/blob/try/watchthemename/gulpfile.babel.js) to help you out.

File: gulpfile.babel.js [hunk 1]
```diff
@@ -111,8 +113,15 @@ function reload(done) {
...
+	// Check if theme slug has been updated.
+	let newThemeSlug = false;
+	if ( config.theme.slug !== themeSlug ) {
+		newThemeSlug = true;
+		themeSlug = config.theme.slug;
+	}
 	return gulp.src(paths.php.src)
-	.pipe(newer(paths.php.dest))
+	// If theme slug has not been updated, run task on changed files only.
+	.pipe(gulpif(!newThemeSlug, newer(paths.php.dest)))
```
Comment (hellofromtonya) [inline gulpfile.babel.js]: We don't need this line of code as it is preventing a rebuild in the following scenario:

1. Files already exist, i.e. where previously built.
2. AND theme's slug and/or name changed in the themeConfig.js file.
3. AND when starting gulp again via `npm run build`.


Removing it fixes the above scenario.  All scenarios do cause the PHP files to rebuild. WooHoo!

File: gulpfile.babel.js [hunk 2]
```diff
@@ -111,8 +113,15 @@ function reload(done) {
  */
 export function php() {
 	config = requireUncached('./dev/config/themeConfig.js');
+	// Check if theme slug has been updated.
```
Comment (hellofromtonya) [inline gulpfile.babel.js]: Lines 116 to 121 can be removed.

File: gulpfile.babel.js [hunk 3]
```diff
@@ -33,11 +33,13 @@ import zip from 'gulp-zip';
 
 // Import theme-specific configurations.
 var config = require('./dev/config/themeConfig.js');
+var themeSlug = config.theme.slug;
```
Comment (hellofromtonya) [inline gulpfile.babel.js]: This line can be removed.

File: gulpfile.babel.js (hunk 1 above)
Comment (mor10) [reply gulpfile.babel.js]: The idea here is for the PHP task to only run on files it needs to run on. The newer method ensures no unnecessary builds. Removing it results in all php files being rebuilt every time a single php file is altered which is resource intensive and unnecessary.

**Comment by mor10** [comment]
Off the top of my head a solution might be to not set `themeSlug` until after the PHP process has run for the first time. That would result in all PHP files being rebuilt every time the build process is started.

File: gulpfile.babel.js (hunk 2 above)
Comment (hellofromtonya) [inline gulpfile.babel.js]: Modify L116-121 to the following code:


```
	let isRebuild = themeConfig.isFirstRun ||
		( themeConfig.slug !== config.theme.slug ) ||
		( themeConfig.name !== config.theme.name );
	if ( isRebuild ) {
		themeConfig.slug = config.theme.slug;
        themeConfig.name = config.theme.name;
	}

	// Reset first run.
	if ( themeConfig.isFirstRun ) {
        themeConfig.isFirstRun = false;
    }
```

File: gulpfile.babel.js (hunk 1 above)
Comment (hellofromtonya) [reply gulpfile.babel.js]: To take into account:

1. On script load, i.e. first run.
2. On name or slug change

we can modify this line of code to be:

```
	// If not a rebuild, then run tasks on changed files only.
	.pipe(gulpif(!isRebuild, newer(paths.php.dest)))
```

File: gulpfile.babel.js (hunk 3 above)
Comment (hellofromtonya) [inline gulpfile.babel.js]: We want to take into account slug and name, as either or both can change.  If either changes, then we want to trigger a rebuild process.  Therefore, I'd change this line of code to:

```
var themeConfig = config.theme;
themeConfig.isFirstRun = true;
```

The advantage here is that we capture the initial theme's configuration.  We'll use this to modify it as we run the php task..

The `themeConfig.isFirstRun` property is capturing the initial "on script load" occurrence.  That way we can run the php function when the first first loads.  It's a cheap way to note that the script is first running.

**Comment by hellofromtonya** [comment]
@mor10 Take a look at this possible solution. Here it's doing the following:

1. Runs the php() task on script load, i.e. 1st run.
2. Runs the rebuild tasks when the theme's slug or name (or both) change.

**Comment by felixarntz** [comment]
General remark: We need to consider whether `slug` will always be the only parameter that needs to be covered here. The slug at the moment is only a single word, however what happens if the the slug would be something consisting of two words? If I want to have `my-theme`, this would only work as a textdomain replacement, but no longer as function name prefix replacement (we'd probably need to sanitize it to `my_theme` there. If the code at some point includes classes or namespaces, we would also need to account for `My_Theme`.

I'm not even sure this issue here is the right one to make that comment, but I generally wanted to highlight this. I wonder whether the starter theme would be better off using distinct `wp_rig` and `wp-rig` strings. That would make replacements in such cases easily possible while at the moment it uses the simplest use-case.

**Comment by mor10** [comment]
@felixarntz can you write that last comment up in a separate issue. I will add it to the 1.1 milestone.

**Pull request merged**